### PR TITLE
fix: store session in workbook

### PR DESCRIPTION
**Pull request by dominique-pfister**

### CODE REVIEW

File: src/OneDrive.js [hunk 1]
```diff
@@ -116,14 +116,6 @@ export class OneDrive {
...
   }
 
-  get workbookSessionId() {
-    return this._sessionId;
-  }
-
-  setWorkbookSessionId(workbookSessionId) {
-    this._sessionId = workbookSessionId;
-  }
-
```
Comment (dominique-pfister) [inline src/OneDrive.js:126]: get rid of global session id

File: src/excel/Workbook.js [hunk 2]
```diff
@@ -23,65 +23,79 @@ export class Workbook extends NamedItemContainer {
...
+  }
+
+  async doFetch(relUrl, rawResponseBody, options) {
+    const opts = { ...options };
+    if (!opts.headers) {
+      opts.headers = {};
+    }
+    if (this._sessionId) {
+      opts.headers['Workbook-Session-Id'] = this._sessionId;
+    }
```
Comment (dominique-pfister) [inline src/excel/Workbook.js:77]: pass session-id if available

File: src/excel/Workbook.js [hunk 3]
```diff
@@ -23,65 +23,79 @@ export class Workbook extends NamedItemContainer {
...
+  }
+
+  async getData() {
+    const result = await this.doFetch(this._uri);
+    return result.value;
+  }
+
+  async getWorksheetNames() {
+    this.log.debug(`get worksheet names from ${this._uri}/worksheets`);
+    const result = await this.doFetch(`${this._uri}/worksheets`);
```
Comment (dominique-pfister) [inline src/excel/Workbook.js:88]: call our own implementation that will add the session id

File: src/excel/Workbook.js [hunk 4]
```diff
@@ -91,20 +105,20 @@ export class Workbook extends NamedItemContainer {
...
   async getTableNames() {
     this.log.debug(`get table names from ${this._uri}/tables`);
-    const result = await this._oneDrive.doFetch(`${this._uri}/tables`);
+    const result = await this.doFetch(`${this._uri}/tables`);
     return result.value.map((v) => v.name);
   }
 
   table(name) {
-    return new Table(this._oneDrive, `${this._uri}/tables`, name, this._log);
+    return new Table(this, `${this._uri}/tables`, name, this._log);
```
Comment (dominique-pfister) [inline src/excel/Workbook.js:121]: pass `us` as doFetch implementor

**Review by tripodsan — CHANGES_REQUESTED** [review]
- ensure the commit will be breaking
- adjust types

**Comment by dominique-pfister** [comment]
> * ensure the commit will be breaking

why is it breaking? the API did not change

> * adjust types

what types?

**Comment by tripodsan** [comment]
> > * ensure the commit will be breaking
> 
> why is it breaking? the API did not change

image

> 
> > * adjust types
> 
> what types?

https://github.com/adobe/helix-onedrive-support/blob/main/src/excel/Workbook.d.ts

but I see that the `OneDrive.getWorkbookSessionId` was never documented.

File: src/excel/Workbook.js [hunk 5]
```diff
@@ -23,65 +23,79 @@ export class Workbook extends NamedItemContainer {
...
-    this.log.debug(`get worksheet names from ${this._uri}/worksheets`);
-    const result = await this._oneDrive.doFetch(`${this._uri}/worksheets`);
-    return result.value.map((v) => v.name);
-  }
-
-  worksheet(name) {
-    return new Worksheet(this._oneDrive, `${this._uri}/worksheets`, name, this._log);
-  }
-
   async createSession() {
```
Comment (tripodsan) [inline src/excel/Workbook.js:26]: this is a bit ugly... I would throw an exception if a session id already exists and add a `getSessionId()`

File: src/excel/Workbook.js [hunk 6]
```diff
@@ -23,65 +23,83 @@ export class Workbook extends NamedItemContainer {
...
     }
+    throw new StatusCodeError('No session associated with workbook', 400);
   }
 
   setSessionId(sessionId) {
-    this._oneDrive.setWorkbookSessionId(sessionId);
+    if (this._sessionId) {
+      throw new StatusCodeError('This workbook is already associated with a session', 400);
+    }
+    this._sessionId = sessionId;
```
Comment (dominique-pfister) [inline src/excel/Workbook.js:71]: reject setting session

File: src/excel/Workbook.js [hunk 7]
```diff
@@ -23,65 +23,83 @@ export class Workbook extends NamedItemContainer {
...
-      return sessionId;
-    } else {
-      const uri = `${this.uri}/createSession`;
-      const result = await this._oneDrive.doFetch(uri, false, {
-        method: 'POST',
-      });
-      this._oneDrive.setWorkbookSessionId(result.id);
-      return result.id;
+    if (this._sessionId) {
+      throw new StatusCodeError('This workbook is already associated with a session', 400);
```
Comment (dominique-pfister) [inline src/excel/Workbook.js:28]: reject overwriting session

File: src/excel/Workbook.js [hunk 8]
```diff
@@ -23,65 +23,87 @@ export class Workbook extends NamedItemContainer {
...
-    } else {
-      throw new StatusCodeError('Please create a session first!', 400);
+      return;
     }
+    throw new StatusCodeError('No session associated with workbook', 400);
+  }
+
+  getSessionId() {
+    return this._sessionId;
   }
```
Comment (dominique-pfister) [inline src/excel/Workbook.js:69]: allow to get session id

**Review by tripodsan — APPROVED** [review]

**Pull request merged**